source/material_model/compositing.cc

```diff
@@ -0,0 +1,219 @@
+                                 "property is not one of the valid values."));
+      return Property::viscosity;
+    }
+
+    //Copy the requested data for one model
+    template <int dim>
+    void
+    Compositing<dim>::composite(const unsigned int model_index,
+                                const typename Interface<dim>::MaterialModelOutputs &evaluated,
+                                typename Interface<dim>::MaterialModelOutputs &out) const
+    {
+      if (model_property_map.at(Property::viscosity) == model_index)
+        out.viscosities = evaluated.viscosities;
+      if (model_property_map.at(Property::density) == model_index)
+        out.densities = evaluated.densities;
+      if (model_property_map.at(Property::thermal_expansion_coefficient) == model_index)
+        out.thermal_expansion_coefficients = evaluated.thermal_expansion_coefficients;
+      if (model_property_map.at(Property::specific_heat) == model_index)
+        out.specific_heat = evaluated.specific_heat;
+      if (model_property_map.at(Property::compressibility) == model_index)
+        out.compressibilities = evaluated.compressibilities;
+      if (model_property_map.at(Property::entropy_derivative_pressure) == model_index)
+        out.entropy_derivative_pressure = evaluated.entropy_derivative_pressure;
+      if (model_property_map.at(Property::entropy_derivative_temperature) == model_index)
+        out.entropy_derivative_temperature = evaluated.entropy_derivative_temperature;
+      if (model_property_map.at(Property::reaction_terms) == model_index)
+        out.reaction_terms == evaluated.reaction_terms;
+    }
+
+    template <int dim>
+    void
+    Compositing<dim>::evaluate(const typename Interface<dim>::MaterialModelInputs &in,
+                               typename Interface<dim>::MaterialModelOutputs &out) const
+    {
+      typename Interface<dim>::MaterialModelOutputs evaluated(out.viscosities.size(),
+                                                              this->introspection().n_compositional_fields);
+
+      for (unsigned int i=0; i<models.size(); ++i)
+        {
+          models[i]->evaluate(in, evaluated);
+          composite(i, evaluated, out);
+        }
+    }
+
+    template <int dim>
+    void
+    Compositing<dim>::declare_parameters (ParameterHandler &prm)
+    {
+      prm.enter_subsection("Material model");
+      {
+        prm.enter_subsection("Compositing");
+        {
+          prm.declare_entry("Models", "",
+                            Patterns::Map(Patterns::Selection("viscosity|density|thermal expansion coefficient|specific heat|compressibility|entropy derivative pressure|entropy derivative temperature|reaction terms"),
+                                          Patterns::Selection(MaterialModel::get_valid_model_names_pattern<dim>())),
+                            "The material property and material model associations used for this composite material model");
```

Comment:
This has been altered to be a list of properties (generated from a static const map listing names and the corresponding enum) which should be much more readable and robust.